Python 3 compatibility issues

[flake8](http://flake8.pycqa.org) testing of https://github.com/mahmoud/boltons on Python 3.7.0

$ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__
```
./misc/table_html_app.py:85:32: F821 undefined name 'basestring'
        if isinstance(context, basestring):  # already serialized
                               ^
./misc/table_html_app.py:96:29: F821 undefined name 'unicode'
            return Response(unicode(context), mimetype="text/plain")
                            ^
./misc/table_html_app.py:118:25: F821 undefined name 'unicode'
        return Response(unicode(context), mimetype="text/plain")
                        ^
./misc/bench_omd.py:43:68: E999 SyntaxError: invalid syntax
    print '(installing werkzeug is recommended for full comparison)'
                                                                   ^
./misc/linkify_changelog.py:47:13: E999 SyntaxError: invalid syntax
    print ret
            ^
./tests/test_ioutils.py:29:17: F821 undefined name 'unicode'
    text_type = unicode
                ^
./boltons/statsutils.py:302:16: F821 undefined name 'mean'
        return mean(self._get_pow_diffs(2))
               ^
./boltons/statsutils.py:324:19: F821 undefined name 'median'
        x = float(median(sorted_vals))  # programmatically defined below
                  ^
./boltons/statsutils.py:325:16: F821 undefined name 'median'
        return median([abs(x - v) for v in sorted_vals])
               ^
./boltons/formatutils.py:312:16: F821 undefined name 'unicode'
        return unicode(self.get_value())
               ^
./boltons/tableutils.py:29:41: F821 undefined name 'unicode'
    string_types, integer_types = (str, unicode), (int, long)
                                        ^
./boltons/tableutils.py:29:57: F821 undefined name 'long'
    string_types, integer_types = (str, unicode), (int, long)
                                                        ^
./boltons/ioutils.py:28:17: F821 undefined name 'unicode'
    text_type = unicode
                ^
./boltons/tbutils.py:31:12: F821 undefined name 'unicode'
    text = unicode
           ^
./boltons/tbutils.py:604:17: F821 undefined name 'unicode'
        value = unicode(value)
                ^
2     E999 SyntaxError: invalid syntax
13    F821 undefined name 'unicode'
15
```

--- a/boltons/formatutils.py
+++ b/boltons/formatutils.py
@@ -41,6 +41,10 @@ from __future__ import print_function
 import re
 from string import Formatter
 
+try:
+    unicode        # Python 2
+except NameError:
+    unicode = str  # Python 3
 
 __all__ = ['DeferredValue', 'get_format_args', 'tokenize_format_str',
            'construct_format_field_str', 'infer_positional_format_args',

--- a/boltons/ioutils.py
+++ b/boltons/ioutils.py
@@ -9,7 +9,6 @@ are useful when dealing with input, output, and bytestreams in a variety of
 ways.
 """
 import os
-import sys
 from io import BytesIO
 from abc import (
     ABCMeta,
@@ -20,13 +19,12 @@ from errno import EINVAL
 from codecs import EncodedFile
 from tempfile import TemporaryFile
 
-# Python2/3 compat
-if sys.version_info[0] == 3:
-    text_type = str
-    binary_type = bytes
-else:
-    text_type = unicode
+try:
+    text_type = unicode  # Python 2
     binary_type = str
+except NameError:
+    text_type = str      # Python 3
+    binary_type = bytes
 
 READ_CHUNK_SIZE = 21333
 """

--- a/boltons/statsutils.py
+++ b/boltons/statsutils.py
@@ -299,6 +299,7 @@ class Stats(object):
         >>> variance(range(97))
         784.0
         """
+        global mean  # defined elsewhere in this file
         return mean(self._get_pow_diffs(2))
     variance = _StatsProperty('variance', _calc_variance)
 
@@ -320,8 +321,9 @@ class Stats(object):
         >>> median_abs_dev(range(97))
         24.0
         """
+        global median  # defined elsewhere in this file
         sorted_vals = sorted(self.data)
-        x = float(median(sorted_vals))  # programmatically defined below
+        x = float(median(sorted_vals))
         return median([abs(x - v) for v in sorted_vals])
     median_abs_dev = _StatsProperty('median_abs_dev', _calc_median_abs_dev)
     mad = median_abs_dev  # convenience

--- a/boltons/tableutils.py
+++ b/boltons/tableutils.py
@@ -27,7 +27,7 @@ from itertools import islice
 from collections import Sequence, Mapping, MutableSequence
 try:
     string_types, integer_types = (str, unicode), (int, long)
-except Exception:
+except NameError:
     # Python 3 compat
     unicode = str
     string_types, integer_types = (str, bytes), (int,)

--- a/boltons/tbutils.py
+++ b/boltons/tbutils.py
@@ -27,10 +27,10 @@ import sys
 import linecache
 
 
-if str is bytes:  # py2
-    text = unicode
-else:  # py3
-    text = str
+try:
+    text = unicode  # Python 2
+except NameError:
+    text = str      # Python 3
 
 
 # TODO: chaining primitives?  what are real use cases where these help?
@@ -601,7 +601,7 @@ def _some_str(value):
     except Exception:
         pass
     try:
-        value = unicode(value)
+        value = text(value)
         return value.encode("ascii", "backslashreplace")
     except Exception:
         pass

--- a/misc/bench_omd.py
+++ b/misc/bench_omd.py
@@ -1,5 +1,6 @@
 # -*- coding: utf-8 -*-
 
+from __future__ import print_function
 import string
 import sys
 sys.path.append('/home/mahmoud/projects/lithoxyl/')
@@ -40,7 +41,7 @@ try:
     from werkzeug.datastructures import MultiDict, OrderedMultiDict as WOMD
     MULTI_IMPLS += (WOMD, MultiDict)
 except ImportError:
-    print '(installing werkzeug is recommended for full comparison)'
+    print('(installing werkzeug is recommended for full comparison)')
 ALL_IMPLS = MULTI_IMPLS + (OD, dict)
 
 
@@ -49,8 +50,8 @@ def bench():
         q_sink = lithoxyl.sinks.QuantileSink()
         impl_name = '.'.join([impl.__module__, impl.__name__])
         log = lithoxyl.logger.BaseLogger(impl_name, sinks=[q_sink])
-        print
-        print '+ %s' % impl_name
+        print()
+        print('+ %s' % impl_name)
         for _ in range(times):
             with log.info('total'):
                 for _ in range(times):
@@ -67,14 +68,14 @@ def bench():
         for action in _all_actions:
             try:
                 best_msecs = q_sink.qas[impl_name][action].min * 1000
-                print '   - %s - %g ms' % (action, best_msecs)
+                print('   - %s - %g ms' % (action, best_msecs))
             except KeyError:
                 pass
         best_msecs = q_sink.qas[impl_name]['total'].min * 1000
         median_msecs = q_sink.qas[impl_name]['total'].median * 1000
-        print ' > ran %d loops of %d items each, best time: %g ms, median time: %g ms' % (times, size, best_msecs, median_msecs)
+        print(' > ran %d loops of %d items each, best time: %g ms, median time: %g ms' % (times, size, best_msecs, median_msecs))
 
-    print
+    print()
     return
 
 

--- a/misc/linkify_changelog.py
+++ b/misc/linkify_changelog.py
@@ -1,3 +1,4 @@
+from __future__ import print_function
 
 import re
 import sys
@@ -44,11 +45,11 @@ def main():
     for (name, url) in sorted(URL_MAP.items()):
         link_map_lines.append('[%s]: %s' % (name, url))
 
-    print ret
-    print
-    print
-    print '\n'.join(link_map_lines)
-    print
+    print(ret)
+    print()
+    print()
+    print('\n'.join(link_map_lines))
+    print()
 
 
 if __name__ == '__main__':

--- a/misc/table_html_app.py
+++ b/misc/table_html_app.py
@@ -19,6 +19,13 @@ _CA_PATH = _CUR_PATH + '/_clastic_assets'
 _CSS_PATH = _CA_PATH + '/common.css'
 _STYLE = open(_CSS_PATH).read()
 
+try:
+    basestring     # Python 2
+    unicode
+except NameError:  # Python 3
+    basestring = (str, )
+    unicode = str
+
 
 def fetch_json(url):
     import urllib2

--- a/tests/test_ioutils.py
+++ b/tests/test_ioutils.py
@@ -1,13 +1,12 @@
 import io
 import os
-import sys
 import codecs
 import random
 import string
 
 try:
     from StringIO import StringIO
-except:
+except ImportError:
     # py3
     StringIO = io.StringIO
 
@@ -21,13 +20,12 @@ from boltons import ioutils
 CUR_FILE_PATH = os.path.abspath(__file__)
 
 
-# Python2/3 compat
-if sys.version_info[0] == 3:
-    text_type = str
-    binary_type = bytes
-else:
-    text_type = unicode
+try:
+    text_type = unicode  # Python 2
     binary_type = str
+except NameError:
+    text_type = str      # Python 3
+    binary_type = bytes
 
 
 class AssertionsMixin(object):

